### PR TITLE
Add pylint and bandit to tox and adapt code to pass linting

### PR DESCRIPTION
**Pull request by lukpueh**
**Description of the changes being introduced by the pull request**:
Adds a lint test environment to tox.ini to run bandit and pylint using securesystemslib's generic (pylintrc)[https://github.com/secure-systems-lab/sample-documents/blob/master/pylintrc].
We separate static from dynamic testing, because former needs to run only once (not for each tested Python version).

The PR also fixes various issues pointed out by running the linters.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by vladimir-v-diaz** [comment]
GitHub indicates that coverage/coveralls failed.  Do you know why?  I took a quick glance but the coveralls output wasn't too helpful.

**Comment by vladimir-v-diaz** [comment]
The Travis output is inundated with test statements.  You should consider printing only test failures to make it easier to review the output and/or any test failures.

**Comment by lukpueh** [comment]
I think it is because the number of covered lines decreased in some files (I removed obsolete imports), which makes not covered lines weigh more and therefor decrease the overall coverage.

**Comment by vladimir-v-diaz** [comment]
What happens if there is a Pylint or Bandit failure?  Does the travis build fail?  Is the pull request blocked if there are any linting failures?

**Comment by vladimir-v-diaz** [comment]
> I think it is because the number of covered lines decreased in some files (I removed obsolete imports), which makes not covered lines weigh more and therefor decrease the overall coverage.

I wonder why GitHub flags the pull request as a failure because of the coveralls check.

File: in_toto/models/layout.py [hunk 1]
```diff
@@ -101,6 +99,11 @@ def __init__(self, **kwargs):
 
     self.validate()
 
+  @property
+  def type_(self):
+    """Getter for protected _type attribute. Trailing underscore used by
+    convention (pep8) to avoid conflict with Python's type keyword. """
```
Comment (vladimir-v-diaz) [inline in_toto/models/layout.py:105]: More importantly, the trailing underscore convention is recommended by our code style guidelines.
https://github.com/secure-systems-lab/code-style-guidelines#function-and-method-arguments

**Comment by lukpueh** [comment]
*Re verbosity:* I added a comment in a related issue and will fix it there https://github.com/in-toto/in-toto/issues/117#issuecomment-355584168.
*Re fail:* If pylint or bandit fail, they return a non-zero value which aborts tox `commands` execution and returns a non-zero exit code, which in turn makes the travis build fail.

**Comment by vladimir-v-diaz** [comment]
> Re fail: If pylint or bandit fail, they return a non-zero value which aborts tox commands execution and returns a non-zero exit code, which in turn makes the travis build fail.

Another thing you should verify is whether the output of a pylint/bandit failure is printed to the console.  If there is a pylint or bandit failure, is it clear from reading Travis's output?

File: in_toto/runlib.py [hunk 2]
```diff
@@ -263,7 +266,7 @@ def execute_link(link_cmd_args, record_streams):
   # btw: we ignore them in the layout anyway
 
   if record_streams:
-    # XXX: Use SpooledTemporaryFile if we expect very large outputs
+    # Use SpooledTemporaryFile if we expect very large outputs
```
Comment (vladimir-v-diaz) [inline in_toto/runlib.py]: Is this still a TODO?  It's not clear from the comment, now that `XXX:` is removed.

**Comment by lukpueh** [comment]
*Re GitHub flag:* There are coveralls settings to configure thresholds, for minimum coverage and maximum coverage decrease.  I see for TUF you set 97% for the former. What to you think is a sensible default for the latter?

**Comment by lukpueh** [comment]
> Another thing you should verify is whether the output of a pylint/bandit failure is printed to the console. If there is a pylint or bandit failure, is it clear from reading Travis's output?

Good catch, I just saw that travis actually didn't run the linters. This is fixed with 3a28ac9.

**Comment by vladimir-v-diaz** [comment]
I'm not sure what's a good number for the maximum coverage decrease.  It is ideal that the majority of pull requests contain some unit testing, and that the minimum test coverage not fall below a certain threshold.

Is setting a minimum threshold below your current coverage sufficient?  Say, 99% minimum test coverage when your current coverage is 100%.  In this way, you can allow a large maximum coverage decrease to avoid any failed coveralls check, and still allow some pull requests that are not fully tested (perhaps it cannot be fully tested because it depends on another pending pull request).

File: in_toto/models/link.py [hunk 3]
```diff
@@ -19,7 +19,6 @@
 """
 
 import attr
-import json
 import securesystemslib.formats
 from in_toto.models.common import Signable
```
Comment (vladimir-v-diaz) [inline in_toto/models/link.py:23]: I see that the "from foo import bar" convention is being used.  Is there a reason why?  This convention is not advised by our guidelines

File: in_toto/runlib.py (hunk 2 above)
Comment (lukpueh) [reply in_toto/runlib.py]: Not really a todo, just a comment to consider this alternative if we run into problems.

File: in_toto/models/link.py (hunk 3 above)
Comment (lukpueh) [reply in_toto/models/link.py:23]: This is not related to this PR. But acknowledged in https://github.com/in-toto/in-toto/issues/145.

File: in_toto/runlib.py (hunk 2 above)
Comment (vladimir-v-diaz) [reply in_toto/runlib.py]: "... to consider this alternative if we run into problems."

Maybe, "We should consider using SpooledTemporaryFile if we expect very large outputs in the future."

**Comment by lukpueh** [comment]
Good thoughts. I set the min coverage threshold to 99% and the max decrease to 0.5%. Latter should ignore fails due to line number decrease. 

In any case, we do allow and have performed merges that fail the coverage/coveralls test.

File: tox.ini [hunk 4]
```diff
@@ -4,9 +4,26 @@
...
+envlist = lint,py27
 
-[testenv]
+[testenv:lint]
+deps =
+    pylint
+    bandit
+
+commands =
+    # Run pylint, using securesystemslib's pylintrc configuration file
```
Comment (vladimir-v-diaz) [inline tox.ini]: pylintrc is provided by "secure systems lab," not "securesystemslib."

File: tox.ini [hunk 5]
```diff
@@ -4,9 +4,26 @@
...
+    # https://github.com/secure-systems-lab/sample-documents/blob/master/pylintrc
+    pylint in_toto
+
+    # Run bandit, a security linter from OpenStack Security
+    # Note: We exclude tests that fail on importing or using the subprocess or
+    # subprocess32 modules, which are considered to have low severity security
+    # implications, but are required by in-toto. We do not skip subprocess
+    # calls with shell=True (B602).
+    bandit -r in_toto --skip B404,B603
```
Comment (vladimir-v-diaz) [inline tox.ini]: Can you link to B404 and B603, to make it easier for the reader to understand the implications of not testing them?

https://docs.openstack.org/bandit/latest/plugins/subprocess_without_shell_equals_true.html

https://docs.openstack.org/bandit/latest/plugins/subprocess_popen_with_shell_equals_true.html

https://docs.openstack.org/bandit/latest/blacklists/blacklist_imports.html#b404-import-subprocess

**Comment by vladimir-v-diaz** [comment]
Okay, that's my final comments for the review.

**Comment by lukpueh** [comment]
Thanks for the review, @vladimir-v-diaz!

**Pull request merged**